O código implementa uma rede neural simples para classificar pacientes com diabetes

In [9]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [10]:
url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv'

#Importanto um DataSet sobre Diabete onde totas as pacientes sao mulheres

df = pd.read_csv(url)

In [11]:
df.head()

6  148  72  35    0  33.6  0.627  50  1
0  1   85  66  29    0  26.6  0.351  31  0
1  8  183  64   0    0  23.3  0.672  32  1
2  1   89  66  23   94  28.1  0.167  21  0
3  0  137  40  35  168  43.1  2.288  33  1
4  5  116  74   0    0  25.6  0.201  30  0

In [12]:
df.columns

Index(['6', '148', '72', '35', '0', '33.6', '0.627', '50', '1'], dtype='object')

In [13]:
#Mudando o nome das colunas
df.columns = [
    'Num Gestasoes',
    'Glicose',
    'Pressao Arterial',
    'Espessura da pele',
    'Insulina',
    'IMC',
    'Funcao Pedigree',
    'Idade',
    'Resultado'
]

In [14]:
df.head()

Num Gestasoes  Glicose  Pressao Arterial  Espessura da pele  Insulina  \
0              1       85                66                 29         0   
1              8      183                64                  0         0   
2              1       89                66                 23        94   
3              0      137                40                 35       168   
4              5      116                74                  0         0   

    IMC  Funcao Pedigree  Idade  Resultado  
0  26.6            0.351     31          0  
1  23.3            0.672     32          1  
2  28.1            0.167     21          0  
3  43.1            2.288     33          1  
4  25.6            0.201     30          0

In [15]:
#Separando meus dados de treinamento

X = df[['Num Gestasoes', 'Glicose', 'Pressao Arterial', 'Espessura da pele', 'Insulina', 'IMC', 'Funcao Pedigree', 'Idade']]
Y = df['Resultado']

In [16]:
from sklearn.model_selection import train_test_split

## Divide os dados em conjuntos de treinamento (80%) e teste (20%)

X_treinamento, X_teste, Y_treinamento, Y_teste = train_test_split(X, Y, test_size=0.2, random_state= 42)

In [17]:
from sklearn.preprocessing import StandardScaler

##Normalizando os dados

scaler = StandardScaler()
X_treinamneto = scaler.fit_transform(X_treinamento)
X_teste = scaler.transform(X_teste)

In [18]:
X.shape

(767, 8)

In [19]:
modelo = Sequential()
modelo.add(Dense(32, activation='relu', input_shape=(8,))) #Espera 8 valores de entrada
modelo.add(Dense(16, activation='relu'))
modelo.add(Dense(8, activation='relu'))
modelo.add(Dense(1, activation='sigmoid')) # saída binária (0 ou 1)


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [20]:
modelo.compile(
    optimizer='adam',
    loss='binary_crossentropy',    # Função para classificação binária
    metrics=['accuracy']
)

In [21]:
modelo.fit(X_treinamento, Y_treinamento, epochs = 100, batch_size=16)

Epoch 1/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.6453 - loss: 5.9100
Epoch 2/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6257 - loss: 0.7819
Epoch 3/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6319 - loss: 0.7360
Epoch 4/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6389 - loss: 0.6814
Epoch 5/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6735 - loss: 0.6878
Epoch 6/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6487 - loss: 0.6436
Epoch 7/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6809 - loss: 0.6239
Epoch 8/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7008 - loss: 0.6101
Epoch 9/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6455 - loss: 0.6649
Epoch 10/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6557 - loss: 0.6157
Epoch 11/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6571 - loss: 0.6034
Epoch 12/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy:

Tinha deixado apenas com uma camda oculta, mais decidi adiconar masi uma na rede e deu uma breve melhorada na acuracci

In [22]:
modelo.evaluate(X_teste, Y_teste) #avaliando desempenho do modelo

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6505 - loss: 0.8661  


[0.9089018702507019, 0.6298701167106628]

Agora vou adicionar um DropOut, para ver se melhora ainda mais

In [23]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

modelo = Sequential()
modelo.add(Dense(32, activation='relu', input_shape=(8,)))
modelo.add(Dropout(0.3))  # desligar 30% dos neurônios dessa camada
modelo.add(Dense(16, activation='relu'))
modelo.add(Dropout(0.3))  # outro dropout
modelo.add(Dense(8, activation='relu'))
modelo.add(Dense(1, activation='sigmoid'))  # camada de saída


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [24]:
modelo.compile(
    optimizer='adam',
    loss='binary_crossentropy',    # Função para classificação binária
    metrics=['accuracy']
)

In [25]:
modelo.fit(X_treinamento, Y_treinamento, epochs = 100, batch_size=16)

Epoch 1/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6204 - loss: 12.9871
Epoch 2/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5328 - loss: 2.3370
Epoch 3/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5396 - loss: 3.0362
Epoch 4/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5570 - loss: 1.6095
Epoch 5/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5297 - loss: 1.2761
Epoch 6/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5723 - loss: 1.2084
Epoch 7/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5453 - loss: 1.1642
Epoch 8/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6406 - loss: 0.8604
Epoch 9/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5959 - loss: 0.8802
Epoch 10/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6450 - loss: 0.8247
Epoch 11/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6101 - loss: 0.8314
Epoch 12/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy

In [ ]:
modelo.evaluate(X_teste, Y_teste) #avaliando desempenho do modelo

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.6505 - loss: 0.6717  


[0.6943615674972534, 0.6298701167106628]

Sem camada extra -> 0.6505	0.7169
Com camada extra -> 0.6583	0.6518
Com camada + Dropout -> 0.6505	0.6717

Acredito que o DropOut funcionou evitanto o Overfitting, por isso a acuraccia abaixou

# Otimizando o modelo

Diminuindo um pouco o DropOut
E aumentando o numero de epocas de treinamento

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

modelo = Sequential()
modelo.add(Dense(32, activation='relu', input_shape=(8,)))
modelo.add(Dropout(0.2))  # desligar 30% dos neurônios dessa camada
modelo.add(Dense(16, activation='relu'))
modelo.add(Dropout(0.2))  # outro dropout
modelo.add(Dense(8, activation='relu'))
modelo.add(Dense(1, activation='sigmoid'))  # camada de saída


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
modelo.compile(
    optimizer='adam',
    loss='binary_crossentropy',    # Função para classificação binária
    metrics=['accuracy']
)

In [ ]:
modelo.fit(X_treinamento, Y_treinamento, epochs = 200, batch_size=16)

Epoch 1/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6552 - loss: 8.2704
Epoch 2/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5745 - loss: 1.9972
Epoch 3/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5520 - loss: 1.2443
Epoch 4/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5814 - loss: 1.1590
Epoch 5/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5506 - loss: 1.0352
Epoch 6/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6011 - loss: 0.9296
Epoch 7/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5888 - loss: 0.8168
Epoch 8/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5619 - loss: 0.7745
Epoch 9/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5934 - loss: 0.7290
Epoch 10/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6401 - loss: 0.8053 
Epoch 11/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6534 - loss: 0.7107
Epoch 12/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy

In [ ]:
modelo.evaluate(X_teste, Y_teste) #avaliando desempenho do modelo

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.6505 - loss: 0.6889 


[0.7138802409172058, 0.6298701167106628]

Perda = 0.713

Acuracia = 0.62

# Fazendo a validaçao cruzada

In [7]:
import pandas as pd
import tensorflow as tf
import numpy as np
import sklearn
import scikeras

In [8]:
from scikeras.wrappers import KerasClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
from tensorflow.keras import backend as k
from tensorflow.keras import utils as np_utils

In [27]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras import backend as k

def criar_rede():
    k.clear_session()

    modelo = Sequential()
    modelo.add(Dense(32, activation='relu', input_shape=(8,)))
    modelo.add(Dropout(0.2))  # Desativa 20% dos neurônios
    modelo.add(Dense(16, activation='relu'))
    modelo.add(Dropout(0.2))  # Outro Dropout de 20%
    modelo.add(Dense(8, activation='relu'))
    modelo.add(Dense(1, activation='sigmoid'))  # Camada de saída para classificação binária

    modelo.compile(optimizer='adam',
                   loss='binary_crossentropy',
                   metrics=['accuracy'])

    return modelo

In [28]:
rede_neural = KerasClassifier(model = criar_rede, epochs=250, batch_size = 10)

In [30]:
resultados = cross_val_score(estimator=rede_neural, X=X ,y=Y, cv=10, scoring="accuracy")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.5779 - loss: 8.4424
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.5531 - loss: 2.3098
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5413 - loss: 1.5265
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5581 - loss: 1.3602
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5734 - loss: 1.0151
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6025 - loss: 1.0214
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6220 - loss: 0.8386
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6481 - loss: 0.8307
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6039 - loss: 0.8075
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6157 - loss: 0.7660
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6388 - loss: 0.7134
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy:

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.5421 - loss: 4.6980
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5661 - loss: 2.9521
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.5645 - loss: 1.6153
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5570 - loss: 1.1299
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5562 - loss: 1.1328
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6164 - loss: 0.8511
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6132 - loss: 0.8701
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6357 - loss: 0.7740
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6313 - loss: 0.7724
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6406 - loss: 0.7320
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6594 - loss: 0.7000
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


69/69 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6089 - loss: 3.1850
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5785 - loss: 1.7871
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5715 - loss: 1.1118
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6178 - loss: 0.9468
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5686 - loss: 0.8391
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5742 - loss: 0.8868
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5542 - loss: 0.7873
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6018 - loss: 0.7591
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6108 - loss: 0.7227
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6128 - loss: 0.7587
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6297 - loss: 0.7082
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6487 - lo

8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
Epoch 1/250


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


69/69 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.4899 - loss: 9.5766
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5351 - loss: 3.4187
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4926 - loss: 2.7060
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5462 - loss: 1.4426
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5499 - loss: 1.0351
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5328 - loss: 0.9529
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6024 - loss: 0.7699
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5879 - loss: 0.8498
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5800 - loss: 0.8331
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6563 - loss: 0.7392
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6126 - loss: 0.7357
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6187 - lo

8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Epoch 1/250


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


69/69 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.4265 - loss: 9.3298
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5486 - loss: 2.7039
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5569 - loss: 1.5863
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5700 - loss: 1.2161
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6109 - loss: 0.9811
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6155 - loss: 0.8662
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5838 - loss: 0.8395
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6347 - loss: 0.7954
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6250 - loss: 0.7728
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6855 - loss: 0.7058
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6246 - loss: 0.7163
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6723 - lo

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


69/69 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.3981 - loss: 7.4029
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5324 - loss: 1.2485
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4990 - loss: 1.1219
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5928 - loss: 0.8328
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6342 - loss: 0.7710
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6136 - loss: 0.7076
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6501 - loss: 0.6909
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6219 - loss: 0.7274
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6514 - loss: 0.7066
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.6563 - loss: 0.6628
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.6837 - loss: 0.6556
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.6648 - lo

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


69/69 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.3805 - loss: 10.4641
Epoch 2/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5326 - loss: 1.8408
Epoch 3/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5148 - loss: 1.6040
Epoch 4/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5255 - loss: 0.9566
Epoch 5/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5504 - loss: 0.9327
Epoch 6/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6070 - loss: 0.7649
Epoch 7/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6698 - loss: 0.6955
Epoch 8/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6899 - loss: 0.6690
Epoch 9/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6698 - loss: 0.6580
Epoch 10/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6662 - loss: 0.7092
Epoch 11/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6322 - loss: 0.6742
Epoch 12/250
69/69 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6470 - l

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


70/70 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.5384 - loss: 8.8460
Epoch 2/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5651 - loss: 4.0824
Epoch 3/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5679 - loss: 1.6705
Epoch 4/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5401 - loss: 1.3311
Epoch 5/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5075 - loss: 1.1017
Epoch 6/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5580 - loss: 1.0907
Epoch 7/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6142 - loss: 0.8759
Epoch 8/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.5866 - loss: 0.9095
Epoch 9/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5982 - loss: 0.6965
Epoch 10/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.6391 - loss: 0.7475
Epoch 11/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5930 - loss: 0.6971
Epoch 12/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5528 - lo

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


70/70 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.3890 - loss: 7.1518
Epoch 2/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4903 - loss: 2.1949
Epoch 3/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5316 - loss: 1.5631
Epoch 4/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5604 - loss: 1.1620
Epoch 5/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5879 - loss: 0.9507
Epoch 6/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5725 - loss: 0.9830
Epoch 7/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6010 - loss: 0.8210
Epoch 8/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6109 - loss: 0.7947
Epoch 9/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6057 - loss: 0.8418
Epoch 10/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6312 - loss: 0.7876
Epoch 11/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5819 - loss: 0.7449
Epoch 12/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6192 - lo

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


70/70 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.5286 - loss: 3.6787
Epoch 2/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6086 - loss: 1.5180
Epoch 3/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.5793 - loss: 0.9348
Epoch 4/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5499 - loss: 0.9278
Epoch 5/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6065 - loss: 0.7444
Epoch 6/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5938 - loss: 0.7617
Epoch 7/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6267 - loss: 0.7492
Epoch 8/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6457 - loss: 0.7384
Epoch 9/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6818 - loss: 0.6897
Epoch 10/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6598 - loss: 0.7077
Epoch 11/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6405 - loss: 0.7009
Epoch 12/250
70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6327 - lo

In [31]:
resultados

array([0.7012987 , 0.71428571, 0.72727273, 0.63636364, 0.72727273,
       0.71428571, 0.75324675, 0.71052632, 0.71052632, 0.71052632])

In [32]:
resultados.mean()

np.float64(0.7105604921394395)

In [33]:
resultados.std()

np.float64(0.028283257829256808)

In [6]:
!pip install scikeras

In [2]:
!pip uninstall scikit-learn -y
!pip install scikit-learn==1.4.2 #Se nao colocar essa versao esta dando incopatibilidadde (dai nao consigo fazer a validaçao cruzada)

Found existing installation: scikit-learn 1.6.1
Uninstalling scikit-learn-1.6.1:
  Successfully uninstalled scikit-learn-1.6.1
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 97.9 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
umap-learn 0.5.9.post2 requires scikit-learn>=1.6, but you have scikit-learn 1.4.2 which is incompatible.


# Fazendo Regressao com os dados de Diabetes



In [36]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras import backend as k
from tensorflow.keras.optimizers import Adam

# Carregar dados
url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv'
colunas = [
    'Num Gestasoes',
    'Glicose',
    'Pressao Arterial',
    'Espessura da pele',
    'Insulina',
    'IMC',
    'Funcao Pedigree',
    'Idade',
    'Resultado'
]
df = pd.read_csv(url, names=colunas)

#Definindo X e Y
y = df['Glicose']
X = df.drop(['Glicose'], axis=1)

# Escalonar
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Dividir
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

#Criando modelo
def criar_rede_regressao():
    k.clear_session()
    modelo = Sequential()
    modelo.add(Dense(32, activation='relu', input_shape=(X.shape[1],)))
    modelo.add(Dense(16, activation='relu'))
    modelo.add(Dense(1, activation='linear'))  # Regressão: ativação linear
    modelo.compile(optimizer=Adam(learning_rate=0.001),
                   loss='mean_squared_error',
                   metrics=['mean_absolute_error'])
    return modelo

# Treinar
modelo = criar_rede_regressao()
modelo.fit(X_treinamento, y_treinamento, epochs=100, batch_size=10, validation_split=0.2)

# Avaliar
loss, erro_medio = modelo.evaluate(X_teste, y_teste)
print(f"Erro médio absoluto (MAE): {erro_medio:.2f}")


Epoch 1/100


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


50/50 ━━━━━━━━━━━━━━━━━━━━ 2s 8ms/step - loss: 15864.7412 - mean_absolute_error: 121.5996 - val_loss: 15670.1270 - val_mean_absolute_error: 120.6941
Epoch 2/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 15084.0938 - mean_absolute_error: 118.6835 - val_loss: 15086.9219 - val_mean_absolute_error: 118.3305
Epoch 3/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 14496.0234 - mean_absolute_error: 116.0037 - val_loss: 13887.4785 - val_mean_absolute_error: 113.3321
Epoch 4/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 12885.8516 - mean_absolute_error: 109.4194 - val_loss: 11812.1660 - val_mean_absolute_error: 104.1134
Epoch 5/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 11023.0029 - mean_absolute_error: 100.5320 - val_loss: 8831.9473 - val_mean_absolute_error: 88.9928
Epoch 6/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 7347.4111 - mean_absolute_error: 81.1918 - val_loss: 5678.1177 - val_mean_absolute_error: 68.7674
Epoch 7/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step -